upf/upf_pfcp_server.c

```diff
@@ -369,9 +369,9 @@ upf_pfcp_server_rx_msg (pfcp_msg_t * msg)
 
 	msg->node = req->node;
 
-	pfcp_msg_pool_put (psm, req);
+	upf_pfcp_handle_msg (req, msg);
```

Comment:
The only real need for passing `req` along is to later extract the session id.

It would be much simpler is you took the `session_index` from req and copied that to `msg`. That would also avoid the need for decoding the request.

The other use case for passing the requests is to ensure that the response message type matches the requests. However, that could be done more generic and for all response types in `upf_pfcp_server_rx_msg` directly.